### PR TITLE
gen: add multigraph power-law generators

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

### CODE REVIEW

File: graph/graphs/gen/batagelj_brandes_test.go [hunk 1]
```diff
@@ -172,3 +174,159 @@ func TestSmallWorldsBBDirected(t *testing.T) {
...
+				t.Errorf("unexpected overlap in partitian membership: n=%d, d=%d: got:%d", n, d, len(o))
+			}
+
+			var minDegree int
+			for _, u := range nodes {
+				var lines int
+				for _, v := range g.From(u) {
+					lines += len(g.Lines(u, v))
+				}
+				if lines > minDegree {
```
Comment (vladimir-ch) [inline graph/graphs/gen/batagelj_brandes_test.go]: And here.

File: graph/graphs/gen/batagelj_brandes.go [hunk 2]
```diff
@@ -289,16 +289,18 @@ func SmallWorldsBB(dst GraphBuilder, n, d int, p float64, src *rand.Rand) error
...
 // Multigraph generators.
 
-type EdgeAdder interface {
-	AddEdge(graph.Edge)
-}
-
-func PreferentialAttachment(dst EdgeAdder, n, d int, src *rand.Rand) {
+// PowerLaw constructs a power-law degree graph by preferential attachment in dst
+// with n nodes and minimum degree d. PowerLaw does not consider nodes in dst prior
+// to the call. PowerLaw does not consider nodes in dst prior to the call.
```
[the remark this answers is not in the document]
Comment (kortschak) [reply graph/graphs/gen/batagelj_brandes.go]: Ugh, I was having buffer issues and I missed all the damage I did.

File: graph/graphs/gen/batagelj_brandes_test.go [hunk 3]
```diff
@@ -172,3 +174,159 @@ func TestSmallWorldsBBDirected(t *testing.T) {
...
+			if len(nodes) != n {
+				t.Errorf("unexpected number of nodes in graph: n=%d, d=%d: got:%d", n, d, len(nodes))
+			}
+			var minDegree int
+			for _, u := range nodes {
+				var lines int
+				for _, v := range g.From(u) {
+					lines += len(g.Lines(u, v))
+				}
+				if lines > minDegree {
```
[the remark this answers is not in the document]
Comment (kortschak) [reply graph/graphs/gen/batagelj_brandes_test.go]: Yeah, I'll just do this properly.

File: graph/graphs/gen/batagelj_brandes_test.go [hunk 4]
```diff
@@ -172,3 +174,159 @@ func TestSmallWorldsBBDirected(t *testing.T) {
...
+			for _, u := range p1 {
+				p1s.Add(u)
+			}
+			p2s := make(set.Nodes)
+			for _, u := range p2 {
+				p2s.Add(u)
+			}
+			o := make(set.Nodes)
+			if o.Intersect(p1s, p2s); len(o) != 0 {
+				t.Errorf("unexpected overlap in partitian membership: n=%d, d=%d: got:%d", n, d, len(o))
```
[the remark this answers is not in the document]
Comment (kortschak) [reply graph/graphs/gen/batagelj_brandes_test.go]: Yes, there are two partitions of the bipartite graph. This checks that they don't share members.

**Comment by kortschak** [comment]
Sorry. PTAL

File: graph/graphs/gen/batagelj_brandes_test.go [hunk 5]
```diff
@@ -172,3 +174,159 @@ func TestSmallWorldsBBDirected(t *testing.T) {
...
+			if len(nodes) != n {
+				t.Errorf("unexpected number of nodes in graph: n=%d, d=%d: got:%d", n, d, len(nodes))
+			}
+			var minDegree int
+			for _, u := range nodes {
+				var lines int
+				for _, v := range g.From(u) {
+					lines += len(g.Lines(u, v))
+				}
+				if lines > minDegree {
```
[the remark this answers is not in the document]
Comment (kortschak) [reply graph/graphs/gen/batagelj_brandes_test.go]: Done

File: graph/graphs/gen/batagelj_brandes_test.go [hunk 6]
```diff
@@ -172,3 +174,159 @@ func TestSmallWorldsBBDirected(t *testing.T) {
...
+				t.Errorf("unexpected overlap in partitian membership: n=%d, d=%d: got:%d", n, d, len(o))
+			}
+
+			var minDegree int
+			for _, u := range nodes {
+				var lines int
+				for _, v := range g.From(u) {
+					lines += len(g.Lines(u, v))
+				}
+				if lines > minDegree {
```
[the remark this answers is not in the document]
Comment (kortschak) [reply graph/graphs/gen/batagelj_brandes_test.go]: Done

File: graph/graphs/gen/batagelj_brandes_test.go (hunk 1 above)
Comment (kortschak) [reply graph/graphs/gen/batagelj_brandes_test.go]: Done

File: graph/graphs/gen/batagelj_brandes_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (vladimir-ch) [reply graph/graphs/gen/batagelj_brandes_test.go]: Sorry, my comments are sometime too minimalist. I was referring to the spelling of "partitian". Should it be "partition"?

File: graph/graphs/gen/batagelj_brandes_test.go [hunk 7]
```diff
@@ -172,3 +174,153 @@ func TestSmallWorldsBBDirected(t *testing.T) {
...
+			for _, u := range p1 {
+				p1s.Add(u)
+			}
+			p2s := make(set.Nodes)
+			for _, u := range p2 {
+				p2s.Add(u)
+			}
+			o := make(set.Nodes)
+			if o.Intersect(p1s, p2s); len(o) != 0 {
+				t.Errorf("unexpected overlap in partitian membership: n=%d, d=%d: got:%d", n, d, len(o))
```
Comment (vladimir-ch) [inline graph/graphs/gen/batagelj_brandes_test.go]: Also here.

**Pull request merged**